### DISCUSSION TITLE
Parsing a numeric string

### DISCUSSION THREAD

**Discussion by imsergiobernal** [colinhacks/zod#330] [Q&A] [answered]
I think that something like this would be awesome

```ts
const numeric = '123'; // typeof string;
const number = zod.string().numeric().int().parse(numeric); // typeof number;
```

numeric() would convert it to a ZodNumber if it was actually a numeric value.

Or maybe, to preserve original string type:
```ts

const numeric = '123'; // typeof string;

const numericString = zod.string().numeric().int().parse(numeric); // typeof string;
const number = zod.number().stringified().int().parse(numeric); // typeof number;
```

What do you think?

**Comment by myywlc** [comment]
i need it

**Comment by colinhacks** [comment]
I'm not a fan of this because it introduces a new method specifically for the string->number conversion. But there are a polynomial number of pairwise conversions (number to string, number to boolean, string to boolean, etc), and introducing a new method for each will get messy. 

More generally, the idea of creating "pipelines" of schemas is is discussed in depth here: https://github.com/colinhacks/zod/issues/264#issuecomment-754051094. In fact, it's how transformers were initially implemented but I moved away from it in Zod 3 for reasons described in that thread. For now, Zod follows a simple "validate, then transform" approach. Validate the data (e.g. check that it's a "numeric string") then convert it into a `number`:

```ts
const numeric = z.string().regex(/^\d+$/).transform(Number);
  console.log(numeric.safeParse("1234")); // => 1234
  console.log(numeric.safeParse("12345.6")); // => ZodError: invalid_string
  console.log(numeric.safeParse("asdf")); // => ZodError: invalid_string
```

**Reply by whengely** [reply]
I've been struggling with this conversion while picking up zod.  I can get the transform down but how can I use the additional methods such as .nonnegative().   It looks like the return of transform is a ZodEffect which doesn't have the typed methods on it.

This is what I want to do but .nonnegative doesn't exist:
`
z.string().transform(x => parseInt(x)).nonnegative()
`

If we're following validate then transform I only have the option to validate the string and access the string methods.

What is the correct way to do this?

**Reply by colinhacks** [reply]
There's no way to do that currently. Zod *validates* the data, then passes the value through a chain of refinements and transforms, that's it. 

Perhaps Zod should probably introduce a "preprocess" method so input can be pre-processed before validation. So you could do something like `z.number().preprocess(val => parseInt(val)).nonnegative()`. But this isn't implemented currently.

**Reply by evantrimboli** [reply]
Something like this would be really useful. Unless I'm missing something, it seems the way to have an optional number field in a form would be to use: `z.string().transform(stringToNumConverter)`. A way to pre-parse the value to convert from string to number (where possible).

For example you could almost have the opposite of a transform:

`z.number().from<string>(s => parseInt(s, 10))`

I'm sure there's a better way, but this is an issue I'm running into at the moment.

**Reply by linde12** [reply]
Was there ever a decision on this? `preprocess` is exactly what i need in order to parse e.g. query parameters coming from a HTTP request.

For example i have a query parameter called `skip` which is a numeric string. I want to say something like:

`skip: z.number().preprocess(s => parseInt(s, 10)).positive().max(100)`

**Reply by scotttrinh** [reply]
@linde12 

[`preprocess` has landed!](https://github.com/colinhacks/zod#preprocess) Here is what it looks like given your example:

```ts
const schema = z.object({
  skip: z.preprocess(
    (a) => parseInt(z.string().parse(a), 10),
    z.number().positive().max(100)
  )
});

**Reply by linde12** [reply]
@scotttrinh Amazing! Thank you!

**Reply by scotttrinh** [reply]
@linde12 

Note that I used `z.string` to coerce the `unknown` to `string` which is what `parseInt` wants, but you could decide to make a type assertion instead to lower your runtime overhead.

```ts
const schema = z.object({
  skip: z.preprocess(
    (a) => parseInt(a as string, 10),
    z.number().positive().max(100)
  )
});

**Reply by linde12** [reply]
@scotttrinh I see, thanks for the tip! I think i'll go with the typecast option. Cheers :1st_place_medal:

**Reply by cyrilchapon** [reply]
@scotttrinh I'm trying to achieve the same but using `.safeParse` on the schema itself.

Either `parseInt(z.string().parse(a), 10)` (which seems legit) or `parseInt(a as string, 10)` can `throw` an Error; which is not gracefully handled by `safeParse` on the generated schema.


Reproduction :
```ts
const schema = z.object({
  skip: z.preprocess(
    (a) => parseInt(z.string().parse(a), 10),
    z.number().positive().max(100)
  )
})

// this throws
schema.safeParse({ skip: ['hey'] })
```

Is there any known elegant way to coerce string along with safe parsing ?

**Reply by kellyrmilligan** [reply]
for me, this is not letting me set a string to the value of the inferred type, am I missing something?

```
skip: z.ZodEffects<z.ZodNumber, number, unknown>
```

is the type I get back. i'm getting the value as a string, and use it in a form as a default value for a field.

**Reply by KeisukeNagakawa** [reply]
how can I validte further for transformed integer?

**Comment by lunacrafts** [comment]
```typescript
import { z, ZodTypeAny } from 'zod';

export const numericString = (schema: ZodTypeAny) => z.preprocess((a) => {
  if (typeof a === 'string') {
    return parseInt(a, 10)
  } else if (typeof a === 'number') {
    return a;
  } else {
    return undefined;
  }
}, schema);


const FindMany = z.object({
  take: numericString(z.number().max(10))
});
```

**Reply by smeijer** [reply]
Any ideas on how to make `z.infer` notice `take` as being a `number`? Currently it comes back as `unknown`.

**Reply by smeijer** [reply]
This cast works, I'm not sure if there's a better way:

```js
const numericString = (schema: z.ZodTypeAny) =>
  z.preprocess((a) => {
    if (typeof a === 'string') {
      return parseInt(a, 10);
    } else if (typeof a === 'number') {
      return a;
    } else {
      return undefined;
    }
  }, schema) as z.ZodEffects<z.ZodTypeAny, number, number>;
  ```

**Reply by Enteleform** [reply]
> This cast works, I'm not sure if there's a better way:

A generic can be used here to allow Zod's default inference to occur:

```ts
import {z, ZodNumber, ZodOptional} from "zod"

function IntegerString
  <schema extends (ZodNumber | ZodOptional<ZodNumber>)>
  (schema:schema)
{
  return (
    z.preprocess((value) => (
      ( (typeof value === "string") ? parseInt(value, 10)
      : (typeof value === "number") ? value
      :                               undefined
    )), schema)
  )
}


type  FindMany = z.infer<typeof FindMany>
const FindMany = z.object({
  take:        IntegerString(z.number().max(10)           ),
  maybeTake:   IntegerString(z.number().max(10).optional()),
  invalidTake: IntegerString(z.string()                   ), // TypeError
})

type take      = FindMany["take"     ] // number
type maybeTake = FindMany["maybeTake"] // number | undefined
```

**Comment by pke** [comment]
I wonder why this is not built in like in joi? Zod is much better with TS support but I always stumble over such things when converting my schemas.

**Reply by kellyrmilligan** [reply]
joi's `number` will take either a string that coerces to number or a number I believe.

**Comment by P4sca1** [comment]
Given the following schema:

```ts
// A number schema, that also accepts strings as input and tries to parse them before validating.
export const numberSchema = z.preprocess((val) => {
	if (typeof val === 'string') return parseInt(val, 10)
	return val
}, z.number())
```

I would expect that it is possible to use number schema methods, e.g. `numberSchema.gte(1)`. Is there any way to make this possible? I just want to validate a number, but it should be possible to also input strings (e.g. for HTTP query params).

**Comment by lgj9172** [comment]
How about this?
```
z.preprocess(
  Number,
  z.number()
),
```

**Reply by pke** [reply]
I have to try this :)

**Reply by lzehrung** [reply]
Beautiful! This style works like a charm defining numeric express route parameters in [zod-express-middleware](https://github.com/Aquila169/zod-express-middleware) and still getting a 'number' property from req.params!

**Reply by iamadetoye** [reply]
Thank you!

**Comment by cyrilchapon** [comment]
For those coming here looking for a a `safeParse`-compatible solution, see my comment in [#696](https://github.com/colinhacks/zod/issues/696#issuecomment-1250912060)

```ts
import { z, ZodType } from 'zod'

const stringToNumberSchema = (def: number) => (z.string().default(`${def}`).transform(Number))
const safePreprocessor = <O, Z extends ZodType<O>> (preprocessorSchema: Z) => (val: unknown): O | null => {
  const parsed = preprocessorSchema.safeParse(val)
  if (!parsed.success) {
    return null
  }
  return parsed.data
}

const _paginationSchema = z.object({
  skip: z.preprocess(
    safePreprocessor(stringToNumberSchema(0)),
    z.number().min(0)
  ),
  limit: z.preprocess(
    safePreprocessor(stringToNumberSchema(20)),
    z.number().min(0).max(100)
  )
}).strict()

export type PaginationQuery = z.infer<typeof _paginationSchema>
```

The default value can also be `undefined` or `null` for various use cases, and it behaves correctly.
I've been using this for several weeks, it seems to do the job

**Comment by Danzo7** [comment]
They should add the support for numericString I think

**Comment by JacobWeisenburger** [comment]
[v3.20](https://github.com/colinhacks/zod/releases/tag/v3.20) just released with [`coerce`](https://github.com/colinhacks/zod#coercion-for-primitives). I think this is what you are looking for.

**Reply by cellulosa** [reply]
Great stuff! I just noticed it doesn't work with enum() and nativeEnum() though

**Reply by JacobWeisenburger** [reply]
Can you give me an example of how you would want to use it with enum?

**Reply by lucasswell** [reply]
Hello Jacob! 

Pretty nice, this new feature made my day. Would be very useful too if would be possible to coerce an array when working with url parameters. For example, a url parameter can be a string or string[] and coerce it to an array the developer could work always as a string[].

**Reply by JacobWeisenburger** [reply]
Is this what you are looking for?

```ts
const schema = z.object({
    string: z.coerce.string(),
    number: z.coerce.number(),
})

console.log(
    schema.parse({
        string: ['string'],
        number: ['123'],
    })
)
// { string: "string", number: 123 }
```

**Reply by lucasswell** [reply]
Inverse of that:

```ts
const schema = z.object({
    string: z.coerce.array(z.string()),
    number: z.coerce.array(z.number()),
})

console.log(
    schema.parse({
        string: 'string',
        number: '123',
    })
)
// { string: ["string"], number: [123] }
```

Because when you are working with url parameters (at least with node express):
- `/?sortBy=name` comes as query parameter `{sortBy: 'name'}`
- `/?sortBy=name&sortBy=age` comes as query parameter `{sortBy: ['name', 'age']}`

And then I have always check if the property `sortBy` is a string or an string[]. It would be great if zod parsed to just string[], returning for the first case `{sortBy: ['name']}`.

**Reply by prscoelho** [reply]
Coercing `''` into number seems to result in 0, since `Number('') = 0` :(

**Reply by JacobWeisenburger** [reply]
Is there a different way you would handle this?

**Reply by metalsadman** [reply]
value coming from a form input coerced to a number still ended up as string upon sending the request even after passing the `zod.parse` :(, I ended up parsing the supposedly number fields again using parseFloat/Int. anyway to solve this atm? ty
```js
// zod object
const params = z.object({
  min: z.coerce.number().min(0),
  max: z.coerce.number().min(0),
  amount: z.coerce.number().min(0),
})

// payload sent
{ amount: "122", max:"33",min: "12" }
```

**Reply by JacobWeisenburger** [reply]
This seems to work as expected for me.
```ts
const params = z.object({
    min: z.coerce.number().min(0),
    max: z.coerce.number().min(0),
    amount: z.coerce.number().min(0),
})

console.log(params.safeParse({ amount: '122', max: '33', min: '12' }))
// { success: true, data: { min: 12, max: 33, amount: 122 } }
```

Can you share a reproducible example?

**Reply by metalsadman** [reply]
hmm, it might be coming from vee-validate, they are using a `toFormValidator` function to prepare the zod schema for their validation logic, the validation passes, but the value remains a string. I'll try posting a repro there.

**Reply by prscoelho** [reply]
> Is there a different way you would handle this?

It was certainly unexpected that `''` resulted a valid number. I don't know if there's anything that should be done, but I know `''` is not a number. I guess it's more an artifact of how js works than anything, I'm just wondering if we could do better and have extra checks beyond Number(input).

**Reply by JacobWeisenburger** [reply]
> I'm just wondering if we could do better and have extra checks beyond Number(input).

What kind of checks would you use for this?

**Reply by prscoelho** [reply]
What do you think about checking for `null` or `''`? Number(input) results in 0 for both of these, while Number(undefined) results in NaN.

**Reply by JacobWeisenburger** [reply]
That could work. How would you implement that in code?

**Reply by prscoelho** [reply]
Nevermind, I read some relevant issues and this seems to be [out of scope](https://github.com/colinhacks/zod/pull/1930#issuecomment-1407794439). Sorry for the noise.

**Reply by cellulosa** [reply]
> Can you give me an example of how you would want to use it with enum?

Mhh something like

z.coerce.boolean().parse("hello").nativeEnum(MyEnum); // => false
z.coerce.boolean().parse("hi").nativeEnum(MyEnum); // => true

**Reply by JacobWeisenburger** [reply]
Is this what you are looking for?

```ts
enum MyEnum { hi = 'hi' }
console.log( z.nativeEnum( MyEnum ).safeParse( 'hi' ).success ) // true
console.log( z.nativeEnum( MyEnum ).safeParse( 'hello' ).success ) // false
```

**Reply by cellulosa** [reply]
> Is this what you are looking for?

it is! Thanks Jacob 🙏 

so perhaps I’m not fully grasping the use of coerce

**Reply by JacobWeisenburger** [reply]
If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by JLarky** [comment, marked as answer]
I think this would be the answer, zod has it's own utility to quickly coerce the type:

```js
console.log(z.coerce.string().parse(1));   // "1"
console.log(z.coerce.string().parse('1')); // "1"
console.log(z.coerce.number().parse(1));   //  1
console.log(z.coerce.number().parse('1')); //  1
```

If you want a bit more control you should look into `preprocess`

```js
console.log(z.preprocess((x) => '' + x, z.string()).parse(1));      // "1"
console.log(z.preprocess((x) => Number(x), z.number()).parse('1')); //  1
```

or `transform` and `pipe`
```js
console.log(z.union([z.string(), z.number()]).transform((x) => '' + x).pipe(z.string()).parse(1));      // "1"
console.log(z.union([z.string(), z.number()]).transform((x) => Number(x)).pipe(z.number()).parse('1')); //  1
```

###### added by @jacobweisenburger

or `pipe` and `coerce`
```ts
console.log(z.union([z.string(), z.number()]).pipe(z.coerce.string()).parse(1));    // "1"
console.log(z.union([z.string(), z.number()]).pipe(z.coerce.number()).parse('1') ); //  1
```

- https://zod.dev/?id=coercion-for-primitives
- https://zod.dev/?id=preprocess
- https://zod.dev/?id=pipe
- https://twitter.com/colinhacks/status/1602238724120924160?lang=en

**Reply by shimon-technonext** [reply]
this does not validate if the incoming value is really a number or not. heres my version of it -

```ts
z.string().refine(v => { let n = Number(v); return !isNaN(n) && v?.length > 0}, {message: "Invalid number"})
```

**Reply by kellyrmilligan** [reply]
There are a lot of powerful solutions on the parsing side, that being said, if you have a form that starts with strings(aka all forms basically?), but need to validate a field is valid for some sort of number situation, i'm not seeing any answer besides a refine based solution ,is that right? otherwise the inferred type will be a number. maybe i'm missing something?

**Reply by fvanwijk** [reply]
You would use refine over coerce when you don't want to lose significant decimals.

**Reply by justinmchase** [reply]
What typescript type will refine be inferred as?

**Reply by clicktodev** [reply]
any updates? i want to validate a string field as a numeric but preserve the type as string.

**Comment by merojosa** [comment]
In case anyone is interested in a different approach:

```js
z.custom<number>()
  .refine((value) => value ?? false, "Required")
  .refine((value) => Number.isFinite(Number(value)), "Invalid number")
  .transform((value) => Number(value));
```

**Comment by zhe** [comment]
My case differs when passing an empty `''` string, expecting `undefined` in safeParse's data and type is `number | undefined`

```typescript
z.coerce.number().optional(); // ❌ get '0' 
```

```typescript
z.preprocess((val) => (val ? val : undefined), z.coerce.number().optional()); // ✅ get 'undefined'
```

**Reply by JacobWeisenburger** [reply]
how about?
```ts
z.coerce.number().nullish().transform( x => x ? x : undefined )
```

**Reply by zhe** [reply]
thanks Jacob, much better. my original answer was causing `unknown`..

`null` is not needed in my case, so i will just use `optional()` instead.

```typescript
z.coerce.number().optional().transform( x => x ? x : undefined )
```

**Reply by zhe** [reply]
just a follow-up:

if narrowing the number down to int 1 - 12, seems `preprocess` only works... 

```ts
// "" -> undefined
// "2" -> pass
// "13" -> fail

// ❌
z.coerce
  .number()
  .int()
  .min(1)
  .max(12)
  .optional()
  .transform((x) => (x ? x : undefined));

// ❌
z.coerce
  .number()
  .optional() // ❌
  .int()
  .min(1)
  .max(12)
  .transform((x) => (x ? x : undefined));

✅
z.preprocess((x) => (x ? x : undefined), z.coerce.number().int().min(1).max(31).optional());

```

**Comment by fallmo** [comment]
Now we can simply do:
```
z.number({ coerce: true })
```

**Reply by pke** [reply]
with a new release?

**Reply by patlux** [reply]
This results into `0` when the string is empty.

**Comment by markedwards** [comment]
It still seems like having a way to indicate a decimal string, with precision, would be very useful.  It's very common to return decimal values as strings, like `"10.3"`.  As it stands, if I'm describing an output schema, for example, all I can here is `z.string()` I think (?).  And in reality the output is highly constrained.

Am I missing something?

**Comment by sahiljain11** [comment]
I just spent a few hours stuck on an downstream issue due to default value of x being 0. The full form of the preprocess should be the following since x ? x : undefined will resolve to undefined when x = 0:

`z.preprocess(x => (x || x === 0 ? x : undefined))`

If anyone else faces a similar issue, hopefully this will resolve it for you :)

**Comment by davidnewcomb** [comment]
I need to parse query strings all the time. These can be undefined, empty strings or numbers like in the example of the page number. That's an hour of my life I'll never get back, so hopefully I can save someone else all that time with this handy one liner.
```
parsePageSchema = z.union([z.string(), z.undefined()]).transform((x) => x ? Number(x) : Number(1)).pipe(z.number().min(1))
parsePageSchema.safeParse(searchParams.page)
```

```
/?page=     -> returns (default) number 1
/?page=2    -> returns number 2
/?page=-3   -> fails !min(1)
/?page=n    -> fails
```

**Reply by hls-app** [reply]
Yes! I am using

```ts
    z
    .string()
    .transform((val) => parseInt(val))
    .pipe(z.number().min(0)),
```

**Comment by OlliL** [comment]
I wonder - what would be the best way to check for a string having a length of 7 chars and only contains numbers? The issue here is, that leading zeros have to be considered. For example "0000001" is valid. Converting it to number first does not work. I tried this first:

```
string().length(7).and(number({coerce: true}))
```
But this leads to "Intersection results could not be merged" when evaluating the and part.
What seems to work is this solution:
```
string().length(7).pipe(number({coerce: true})),
```
But I wonder - are there any better solutions I could use to face the problem?

**Reply by grundb** [reply]
You don't need zod for this. Just `/^\d{7}$/.test(s)` to test if `s` matches your conditions. If you want zod you could also do `z.string().regex(/^\d{7}$/).transform(Number)`

**Comment by almogzur** [comment]
```
  price: z
     .string()
    .regex(/^(?:[1-9]\d*|0)(?:\.\d{1,2})?$/g,{message:"not valid number"})
    .or(z.number().nonnegative().min(1))
```
this work for me

**Comment by rnnc** [comment]
you can use a custom type and transform for a more elegant reusable solution

```ts
const numstr_regex = /^[+-]?(\d*|\d{1,3}(,\d{3})*)(\.\d+)?\b$/;

export const zNumStr = z.custom<number | string>(
  (val) => (typeof val === 'string') ? numstr_regex.test(val) : (typeof val === 'number'),
  'invalid number'
).transform(v => typeof v === 'string' ? Number(v.replace(',', '')) : v);
```

^ this works for strings like "12,345.678"

but if you don't care/want commas to be supported, use this regex & change the transform

```ts
const numstr_regex = /^[+-]?(\d*\.)?\d+$/;
// ....
.transform(v typeof v === 'string' ? Number(v) : v);
```

and if you want the parsed result to be a string instead you can either remove the transform, or just add this custom transform on top of the zod type for ad-hoc usage

```ts
const parsed = zNumStr.transform(v => `${v}`).parse(value);
```

**Comment by clicktodev** [comment]
I ended up using this for strings that represent int64

if there's a better approach, let me know

```ts
import { z } from "zod";

export const int64String = z.string().refine((val) => {
  try {
    // Let zod do the int64 check
    z.int64().parse(BigInt(val));
    return true;
  } catch {
    return false;
  }
  }, {
    message: "Must be a valid int64 string",
  });

int64String.parse("123");                      // OK
int64String.parse("9223372036854775807");     // OK
int64String.parse("9223372036854775808");     // ❌ outside int64
int64String.parse("abc");                     // ❌ invalid

```

**Comment by bloodykheeng** [comment]
i think a easier approach is to  specify the input type generic
ie number<string>()

see code below
```
quantity: z.coerce.number<number>().min(1, "Quantity must be at least 1"),
total_price: z.coerce.number<number>().nullish()
```

if its not specified react hook form will try to infer and see unknown because of coerce but it the type generic is specified the 
it will see number
